core/src/main/scala/io/projectglow/sql/expressions/LinearRegressionGwas.scala

```diff
@@ -32,173 +30,48 @@ object LinearRegressionGwas extends GlowLogging {
-      covariateQRContext.tau,
-      genotypes,
-      genotypes.length,
-      covariateQRContext.work,
-      covariateQRContext.work.length,
-      out
-    )
-
-    require(
-      out.`val` == 0,
-      s"Computing Q_C ^T * g because argument ${-out.`val`} had illegal value")
-
-    // Zero out elements below the diagonal of Q_C^T * g with a Householder transformation
-    val alpha = new doubleW(genotypes(covariateQRContext.numCovariateCols))
-    val v = new Array[Double](genotypes.length - covariateQRContext.numCols)
-    val lastTau = new doubleW(0)
-    System.arraycopy(genotypes, covariateQRContext.numCols, v, 0, v.length)
-    lapack.dlarfg(v.length + 1, alpha, v, 1, lastTau)
-
-    // We now have the complete final column of QR_G. Copy it into the QR array.
-    System.arraycopy(
-      genotypes,
-      0,
-      covariateQRContext.values,
-      initialNumElements,
-      covariateQRContext.numCovariateCols)
-    covariateQRContext.values(initialNumElements + covariateQRContext.numCovariateCols) =
-      alpha.`val`
-    System.arraycopy(
-      v,
-      0,
-      covariateQRContext.values,
-      initialNumElements + covariateQRContext.numCols,
-      v.length)
-
-    // Add the final tau
-    covariateQRContext.tau(covariateQRContext.tau.length - 1) = lastTau.`val`
-
-    // Now covariateQRContext contains QR_G
-
-    // Get d = Q_G^T * b
-    lapack.dormqr(
-      "L",
-      "T",
-      phenotypes.length,
-      1,
-      covariateQRContext.tau.length,
-      covariateQRContext.values,
-      phenotypes.length,
-      covariateQRContext.tau,
-      phenotypes,
-      phenotypes.length,
-      covariateQRContext.work,
-      covariateQRContext.work.length,
-      out
-    )
-
-    require(out.`val` == 0, s"Computing d failed because argument ${-out.`val`} had illegal value")
-
-    // Now phenotypes holds d, so we solve for x (the regression coefficients)
-    lapack.dtrtrs(
-      "U",
-      "N",
-      "N",
-      covariateQRContext.numCols,
-      1,
-      covariateQRContext.values,
-      covariateQRContext.numRows,
-      phenotypes,
-      phenotypes.length,
-      out)
-
-    require(
-      out.`val` >= 0,
-      s"Solving for beta failed because argument ${-out.`val`} had illegal value")
-    require(out.`val` == 0, s"Solving for beta failed because G was singular")
-
-    val beta = phenotypes(covariateQRContext.numCovariateCols)
-
-    // The total error is the d^T * d for the rows greater than (n_covariates + 1)
-    val totalError = blas.ddot(
-      phenotypes.length - covariateQRContext.numCols,
-      phenotypes,
-      covariateQRContext.numCols,
-      1,
-      phenotypes,
-      covariateQRContext.numCols,
-      1
-    )
-
-    // Calculate the t statistic described in
-    // https://web.stanford.edu/~hastie/ElemStatLearn/printings/ESLII_print12.pdf equation
-    // 3.12, we need to invert (G^T * G)
-    // Note that (G^T * G) = (R_G^T * R_G)
-    val qrCopy = covariateQRContext.values.clone()
```

Comment:
To clarify, were these two lines the expensive ones?

Reply:
I think so. I couldn't actually see this is in the profile directly, but i believe that this led to more GC and worse cache locality.